diff --git a/.gitmodules b/.gitmodules
index 36d5becb57c..ab10795624c 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -55,7 +55,6 @@
     path = third_party/onnx
     url = https://github.com/onnx/onnx.git
 [submodule "third_party/sleef"]
-    ignore = dirty
     path = third_party/sleef
     url = https://github.com/shibatch/sleef
 [submodule "third_party/ideep"]
diff --git a/CMakeLists.txt b/CMakeLists.txt
index c8af5f00b5c..4d3b177b2fe 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -22,6 +22,16 @@ if(${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
   message(FATAL_ERROR "In-source build are not supported")
 endif()
 
+# --------------------------------- chaquopy begin ------------------------------------
+set(BUILD_TEST OFF)
+set(USE_CUDA OFF)
+set(USE_OPENCV OFF)
+set(USE_VULKAN OFF)
+set(BUILD_PYTHON ON)
+set(Python_EXECUTABLE "/home/ubuntu/workspace/chaquopy/server/pypi/packages/torch/build/2.6.0/cp39-cp39-android_28_arm64_v8a/env/bin/python3.9" CACHE STRING "Python interpreter" FORCE)
+set(Python_NumPy_INCLUDE_DIRS "/home/ubuntu/workspace/chaquopy/server/pypi/packages/torch/build/2.6.0/cp39-cp39-android_28_arm64_v8a/requirements/numpy/core/include" CACHE STRING "NumPy include dir" FORCE)
+# --------------------------------- chaquopy end ------------------------------------
+
 # ---[ Project and semantic versioning.
 project(Torch CXX C)
 
@@ -663,9 +673,9 @@ string(APPEND CMAKE_CUDA_FLAGS " -Xfatbin -compress-all")
 if(ANDROID
    OR IOS
    OR DEFINED ENV{BUILD_PYTORCH_MOBILE_WITH_HOST_TOOLCHAIN})
-  set(INTERN_BUILD_MOBILE ON)
+  # set(INTERN_BUILD_MOBILE ON)
   message(WARNING "INTERN_BUILD_MOBILE is on, disabling BUILD_LAZY_TS_BACKEND")
-  set(BUILD_LAZY_TS_BACKEND OFF)
+  # set(BUILD_LAZY_TS_BACKEND OFF)
 
   # Set -ffunction-sections and -fdata-sections so that each method has its own
   # text section. This allows the linker to remove unused section when the flag
diff --git a/c10/macros/Macros.h b/c10/macros/Macros.h
index 919eb6c8567..b878027c648 100644
--- a/c10/macros/Macros.h
+++ b/c10/macros/Macros.h
@@ -433,7 +433,8 @@ __host__ __device__
 
 #if defined(__ANDROID__)
 #define C10_ANDROID 1
-#define C10_MOBILE 1
+// Chaquopy: this disables things required by the Python build.
+// #define C10_MOBILE 1
 #elif (                   \
     defined(__APPLE__) && \
     (TARGET_IPHONE_SIMULATOR || TARGET_OS_SIMULATOR || TARGET_OS_IPHONE))
diff --git a/cmake/Dependencies.cmake b/cmake/Dependencies.cmake
index 1813f4418a2..d55e7551ee7 100644
--- a/cmake/Dependencies.cmake
+++ b/cmake/Dependencies.cmake
@@ -848,7 +848,13 @@ if(BUILD_PYTHON)
   if(USE_NUMPY)
     list(APPEND PYTHON_COMPONENTS NumPy)
   endif()
-  find_package(Python COMPONENTS Interpreter OPTIONAL_COMPONENTS ${PYTHON_COMPONENTS})
+  # Chaquopy
+  # find_package(Python COMPONENTS Interpreter OPTIONAL_COMPONENTS ${PYTHON_COMPONENTS})
+  # set(Python_NumPy_INCLUDE_DIRS "/home/ubuntu/workspace/chaquopy/server/pypi/packages/torch/build/2.6.0/cp39-cp39-android_28_arm64_v8a/requirements/numpy/core/include")
+  # include_directories(${Python_NumPy_INCLUDE_DIRS})
+  message(STATUS "Python_NumPy_INCLUDE_DIRS = ${Python_NumPy_INCLUDE_DIRS}")
+  find_package(NumPy REQUIRED)
+  find_package(Python COMPONENTS Interpreter OPTIONAL_COMPONENTS Development ${PYTHON_COMPONENTS})
 else()
   find_package(Python COMPONENTS Interpreter)
 endif()
@@ -864,8 +870,11 @@ endif()
 
 # ---[ Python + Numpy
 if(BUILD_PYTHON)
+  message(----------------------------Python_Development.Module_FOUND = ${Python_Development.Module_FOUND})
   if(Python_Development.Module_FOUND)
+    message(----------------------------USE_NUMPY = ${USE_NUMPY})
     if(USE_NUMPY)
+    message(----------------------------Python_NumPy_FOUND = ${Python_NumPy_FOUND})
       if(NOT Python_NumPy_FOUND)
         message(WARNING "NumPy could not be found. Not building with NumPy. Suppress this warning with -DUSE_NUMPY=OFF")
         caffe2_update_option(USE_NUMPY OFF)
@@ -880,6 +889,9 @@ if(BUILD_PYTHON)
     caffe2_update_option(BUILD_PYTHON OFF)
   endif()
 endif()
+message(----------------------------1---USE_NUMPY = ${USE_NUMPY})
+caffe2_update_option(USE_NUMPY ON)
+message(----------------------------2---USE_NUMPY = ${USE_NUMPY})
 
 # ---[ pybind11
 if(USE_SYSTEM_PYBIND11)
@@ -1290,7 +1302,8 @@ if(CAFFE2_CMAKE_BUILDING_WITH_MAIN_REPO AND NOT INTERN_DISABLE_ONNX)
     # In mobile build we care about code size, and so we need drop
     # everything (e.g. checker) in onnx but the pb definition.
     if(ANDROID OR IOS)
-      caffe2_interface_library(onnx_proto onnx_library)
+      # caffe2_interface_library(onnx_proto onnx_library)
+      caffe2_interface_library(onnx onnx_library)  # Chaquopy: `onnx` was `onnx_proto`.
     else()
       caffe2_interface_library(onnx onnx_library)
     endif()
@@ -1497,7 +1510,9 @@ if(NOT INTERN_BUILD_MOBILE)
       add_definitions(-DHAVE_MMAP=1)
     endif(HAVE_MMAP)
     # done for lseek: https://www.gnu.org/software/libc/manual/html_node/File-Position-Primitive.html
-    add_definitions(-D_FILE_OFFSET_BITS=64)
+
+    # Chaquopy: see https://android.googlesource.com/platform/bionic/+/master/docs/32-bit-abi.md
+    # add_definitions(-D_FILE_OFFSET_BITS=64)
     CHECK_FUNCTION_EXISTS(shm_open HAVE_SHM_OPEN)
     if(HAVE_SHM_OPEN)
       add_definitions(-DHAVE_SHM_OPEN=1)
diff --git a/functorch/CMakeLists.txt b/functorch/CMakeLists.txt
index bdfa4bfe455..9e3b4643097 100644
--- a/functorch/CMakeLists.txt
+++ b/functorch/CMakeLists.txt
@@ -22,7 +22,7 @@ else()
   # functorch cannot use the alias to build on windows
   target_link_libraries(${PROJECT_NAME} PRIVATE torch torch_python)
 endif()
-target_link_libraries(${PROJECT_NAME} PRIVATE pybind::pybind11)
+target_link_libraries(${PROJECT_NAME} PRIVATE pybind::pybind11 ${PYTHON_LIBRARIES})
 
 set_target_properties(${PROJECT_NAME} PROPERTIES LIBRARY_OUTPUT_DIRECTORY
       ${CMAKE_BINARY_DIR}/functorch)
diff --git a/setup.py b/setup.py
index a6a6db7b033..6a9a09d8a43 100644
--- a/setup.py
+++ b/setup.py
@@ -368,7 +368,9 @@ if BUILD_LIBTORCH_WHL:
 
 
 package_type = os.getenv("PACKAGE_TYPE", "wheel")
-version = get_torch_version()
+# todo xiao
+# version = get_torch_version()
+version = "2.6.0"
 report(f"Building wheel {package_name}-{version}")
 
 cmake = CMake()
@@ -487,7 +489,7 @@ def build_deps():
         version=version,
         cmake_python_library=cmake_python_library,
         build_python=build_python,
-        rerun_cmake=RERUN_CMAKE,
+        rerun_cmake=True,
         cmake_only=CMAKE_ONLY,
         cmake=cmake,
     )
@@ -1167,7 +1169,15 @@ def main():
         long_description = f.read()
 
     version_range_max = max(sys.version_info[1], 13) + 1
+    
+    # Chaquopy: see package_data below.
+    if os.path.exists("chaquopy"):
+        shutil.rmtree("chaquopy")
+    os.mkdir("chaquopy")
+    os.symlink("../torch/lib", "chaquopy/lib")
+    
     torch_package_data = [
+        'lib/libtorch_global_deps.so',  # Chaquopy: see comment at top of package_data.
         "py.typed",
         "bin/*",
         "test/*",
@@ -1416,6 +1426,9 @@ def main():
     ]
     package_data = {
         "torch": torch_package_data,
+        'chaquopy': [
+            'lib/*.so*'  # See exclude_package_data below.
+        ],
     }
 
     if not BUILD_LIBTORCH_WHL:
@@ -1435,7 +1448,7 @@ def main():
         long_description_content_type="text/markdown",
         ext_modules=extensions,
         cmdclass=cmdclass,
-        packages=packages,
+        packages=packages + ["chaquopy"],
         entry_points=entry_points,
         install_requires=install_requires,
         extras_require=extras_require,
@@ -1445,6 +1458,9 @@ def main():
         author="PyTorch Team",
         author_email="packages@pytorch.org",
         python_requires=f">={python_min_version_str}",
+        
+        exclude_package_data={'chaquopy': ['lib/libtorch_global_deps.so']},
+        
         # PyPI package information.
         classifiers=[
             "Development Status :: 5 - Production/Stable",
Submodule third_party/sleef contains modified content
diff --git a/third_party/sleef/CMakeLists.txt b/third_party/sleef/CMakeLists.txt
index ad31386..59b102a 100644
--- a/third_party/sleef/CMakeLists.txt
+++ b/third_party/sleef/CMakeLists.txt
@@ -12,7 +12,7 @@ option(SLEEF_BUILD_DFT "libsleefdft will be built." OFF)
 option(SLEEF_BUILD_QUAD "libsleefquad will be built." OFF)
 option(SLEEF_BUILD_GNUABI_LIBS "libsleefgnuabi will be built." ON)
 option(SLEEF_BUILD_SCALAR_LIB "libsleefscalar will be built." OFF)
-option(SLEEF_BUILD_TESTS "Tests will be built." ON)
+option(SLEEF_BUILD_TESTS "Tests will be built." OFF)
 option(SLEEF_BUILD_INLINE_HEADERS "Build header for inlining whole SLEEF functions" OFF)
 
 option(SLEEF_TEST_ALL_IUT "Perform tests on implementations with all vector extensions" OFF)
diff --git a/third_party/sleef/Configure.cmake b/third_party/sleef/Configure.cmake
index 19a153f..bcb96f6 100644
--- a/third_party/sleef/Configure.cmake
+++ b/third_party/sleef/Configure.cmake
@@ -115,8 +115,10 @@ elseif(SLEEF_TARGET_PROCESSOR MATCHES "aarch64|arm64")
 
 elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "arm")
   set(SLEEF_ARCH_AARCH32 ON CACHE INTERNAL "True for Aarch32 architecture.")
-  set(COMPILER_SUPPORTS_NEON32 1)
-  set(COMPILER_SUPPORTS_NEON32VFPV4 1)
+  # set(COMPILER_SUPPORTS_NEON32 1)
+  # set(COMPILER_SUPPORTS_NEON32VFPV4 1)
+  set(COMPILER_SUPPORTS_NEON32 0)  # Chaquopy: changed from 1 to 0
+  set(COMPILER_SUPPORTS_NEON32VFPV4 0)  # Chaquopy: changed from 1 to 0
 
   set(CLANG_FLAGS_ENABLE_PURECFMA_SCALAR "-mfpu=vfpv4")
 elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc|ppc)64")
@@ -133,7 +135,8 @@ elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "riscv64")
 endif()
 
 set(COMPILER_SUPPORTS_PUREC_SCALAR 1)
-set(COMPILER_SUPPORTS_PURECFMA_SCALAR 1)
+# set(COMPILER_SUPPORTS_PURECFMA_SCALAR 1)
+set(COMPILER_SUPPORTS_PURECFMA_SCALAR 0)  # Chaquopy: changed from 1 to 0
 
 # Compiler feature detection
 
diff --git a/tools/setup_helpers/cmake.py b/tools/setup_helpers/cmake.py
index 84e4dad32d3..a4434cb9974 100644
--- a/tools/setup_helpers/cmake.py
+++ b/tools/setup_helpers/cmake.py
@@ -6,6 +6,7 @@ import multiprocessing
 import os
 import platform
 import sys
+import subprocess
 import sysconfig
 from distutils.version import LooseVersion
 from subprocess import CalledProcessError, check_call, check_output
@@ -305,10 +306,27 @@ class CMake:
             )
             sys.exit(1)
         build_options.update(cmake__options)
-
+        
+        # Chaquopy: see https://github.com/shibatch/sleef/issues/249
+        build_sleef_native = os.path.abspath("build-sleef-native")
+        if not os.path.exists(build_sleef_native):  # For rerunning with build-wheel.py --no-unpack
+            os.mkdir(build_sleef_native)
+            sleef_env = os.environ.copy()
+            for name in ["AR", "ARFLAGS", "AS", "CC", "CFLAGS", "CPP", "CPPFLAGS", "CXX",
+                        "CXXFLAGS", "F77", "F90", "FARCH", "FC", "LD", "LDFLAGS", "LDSHARED",
+                        "NM", "RANLIB", "READELF", "STRIP", "CMAKE_TOOLCHAIN_FILE"]:
+                sleef_env.pop(name, None)
+            kwargs = dict(cwd=build_sleef_native, env=sleef_env)
+            check_call(["cmake", "../third_party/sleef"], **kwargs)
+            check_call(["cmake", "--build", "." , "--", "-j", str(os.cpu_count())], **kwargs)
+        
         CMake.defines(
             args,
-            Python_EXECUTABLE=sys.executable,
+            # Python_EXECUTABLE=sys.executable,
+            CAFFE2_CUSTOM_PROTOC_EXECUTABLE=which("protoc"),
+            ONNX_CUSTOM_PROTOC_EXECUTABLE=which("protoc"),
+            # BLAS="OpenBLAS",
+            NATIVE_BUILD_DIR=build_sleef_native,
             TORCH_BUILD_VERSION=version,
             **build_options,
         )
diff --git a/torch/CMakeLists.txt b/torch/CMakeLists.txt
index b123023d2fd..4ebd88274ae 100644
--- a/torch/CMakeLists.txt
+++ b/torch/CMakeLists.txt
@@ -428,7 +428,7 @@ endif()
 
 target_compile_definitions(torch_python PRIVATE "-DTHP_BUILD_MAIN_LIB")
 
-target_link_libraries(torch_python PRIVATE ${TORCH_LIB} ${TORCH_PYTHON_LINK_LIBRARIES})
+target_link_libraries(torch_python PRIVATE ${TORCH_LIB} ${TORCH_PYTHON_LINK_LIBRARIES} ${PYTHON_LIBRARIES})
 
 target_compile_definitions(torch_python PRIVATE ${TORCH_PYTHON_COMPILE_DEFINITIONS})
 
@@ -492,7 +492,7 @@ if(NOT ${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
   if(BUILD_LIBTORCHLESS)
     target_link_libraries(nnapi_backend PRIVATE ${TORCH_LIB} torch_python pybind::pybind11)
   else()
-    target_link_libraries(nnapi_backend PRIVATE torch torch_python pybind::pybind11)
+    target_link_libraries(nnapi_backend PRIVATE torch torch_python pybind::pybind11 ${PYTHON_LIBRARIES})
   endif()
 endif()
 
diff --git a/torch/_utils_internal.py b/torch/_utils_internal.py
index f58eb93d86d..b72d68cf59b 100644
--- a/torch/_utils_internal.py
+++ b/torch/_utils_internal.py
@@ -63,7 +63,8 @@ def prepare_multiprocessing_environment(path: str) -> None:
 
 
 def resolve_library_path(path: str) -> str:
-    return os.path.realpath(path)
+    # return os.path.realpath(path)
+    return os.path.abspath(path)  # Chaquopy: was realpath. This patch is no longer necessary on newer versions of Chaquopy, which call realpath early in the importer.
 
 
 def throw_abstract_impl_not_imported_error(opname, module, context):
diff --git a/torch/csrc/inductor/aoti_eager/kernel_holder.cpp b/torch/csrc/inductor/aoti_eager/kernel_holder.cpp
index 6b2ee88df9c..8ca22541dec 100644
--- a/torch/csrc/inductor/aoti_eager/kernel_holder.cpp
+++ b/torch/csrc/inductor/aoti_eager/kernel_holder.cpp
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #include <torch/csrc/inductor/aoti_eager/kernel_holder.h>
 
 #include <ATen/ATen.h>
diff --git a/torch/csrc/inductor/aoti_eager/kernel_holder.h b/torch/csrc/inductor/aoti_eager/kernel_holder.h
index fed2e3b5d61..0b28c4f2852 100644
--- a/torch/csrc/inductor/aoti_eager/kernel_holder.h
+++ b/torch/csrc/inductor/aoti_eager/kernel_holder.h
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #pragma once
 
 #include <ATen/ATen.h>
diff --git a/torch/csrc/inductor/aoti_eager/kernel_meta_info.cpp b/torch/csrc/inductor/aoti_eager/kernel_meta_info.cpp
index 1642ee4beca..ba0a022843b 100644
--- a/torch/csrc/inductor/aoti_eager/kernel_meta_info.cpp
+++ b/torch/csrc/inductor/aoti_eager/kernel_meta_info.cpp
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #include <torch/csrc/inductor/aoti_eager/kernel_meta_info.h>
 #include <iostream>
 #include <utility>
diff --git a/torch/csrc/inductor/aoti_eager/kernel_meta_info.h b/torch/csrc/inductor/aoti_eager/kernel_meta_info.h
index 24d3c05bc35..d192e11a836 100644
--- a/torch/csrc/inductor/aoti_eager/kernel_meta_info.h
+++ b/torch/csrc/inductor/aoti_eager/kernel_meta_info.h
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #pragma once
 
 #include <ATen/ATen.h>
diff --git a/torch/csrc/inductor/aoti_package/model_package_loader.cpp b/torch/csrc/inductor/aoti_package/model_package_loader.cpp
index 1aa408bcafd..cc516609e21 100644
--- a/torch/csrc/inductor/aoti_package/model_package_loader.cpp
+++ b/torch/csrc/inductor/aoti_package/model_package_loader.cpp
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 
 #include <c10/util/error.h>
 #include <torch/csrc/inductor/aoti_package/model_package_loader.h>
diff --git a/torch/csrc/inductor/aoti_package/model_package_loader.h b/torch/csrc/inductor/aoti_package/model_package_loader.h
index 88be54ea4f2..62958b79614 100644
--- a/torch/csrc/inductor/aoti_package/model_package_loader.h
+++ b/torch/csrc/inductor/aoti_package/model_package_loader.h
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #pragma once
 
 #include <ATen/Tensor.h>
diff --git a/torch/csrc/inductor/aoti_runner/model_container_runner.cpp b/torch/csrc/inductor/aoti_runner/model_container_runner.cpp
index 41dfd588d48..a943acd1a9a 100644
--- a/torch/csrc/inductor/aoti_runner/model_container_runner.cpp
+++ b/torch/csrc/inductor/aoti_runner/model_container_runner.cpp
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #include <ATen/DynamicLibrary.h>
 
 #include <torch/csrc/inductor/aoti_runner/model_container_runner.h>
diff --git a/torch/csrc/inductor/aoti_runner/model_container_runner.h b/torch/csrc/inductor/aoti_runner/model_container_runner.h
index 76bda78ff23..6bd04331235 100644
--- a/torch/csrc/inductor/aoti_runner/model_container_runner.h
+++ b/torch/csrc/inductor/aoti_runner/model_container_runner.h
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #pragma once
 
 #include <ATen/Tensor.h>
diff --git a/torch/csrc/inductor/aoti_runner/model_container_runner_cpu.cpp b/torch/csrc/inductor/aoti_runner/model_container_runner_cpu.cpp
index ecae588be2f..702d9467508 100644
--- a/torch/csrc/inductor/aoti_runner/model_container_runner_cpu.cpp
+++ b/torch/csrc/inductor/aoti_runner/model_container_runner_cpu.cpp
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #include <torch/csrc/inductor/aoti_runner/model_container_runner_cpu.h>
 
 namespace torch::inductor {
diff --git a/torch/csrc/inductor/aoti_runner/model_container_runner_cpu.h b/torch/csrc/inductor/aoti_runner/model_container_runner_cpu.h
index 7a38dab5c2f..fc8ac7f2312 100644
--- a/torch/csrc/inductor/aoti_runner/model_container_runner_cpu.h
+++ b/torch/csrc/inductor/aoti_runner/model_container_runner_cpu.h
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #pragma once
 
 #include <torch/csrc/inductor/aoti_runner/model_container_runner.h>
diff --git a/torch/csrc/inductor/aoti_runner/model_container_runner_cuda.cpp b/torch/csrc/inductor/aoti_runner/model_container_runner_cuda.cpp
index c5abd97f759..9e7f0418d92 100644
--- a/torch/csrc/inductor/aoti_runner/model_container_runner_cuda.cpp
+++ b/torch/csrc/inductor/aoti_runner/model_container_runner_cuda.cpp
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #include <torch/csrc/inductor/aoti_runner/model_container_runner_cuda.h>
 
 namespace torch::inductor {
diff --git a/torch/csrc/inductor/aoti_runner/model_container_runner_cuda.h b/torch/csrc/inductor/aoti_runner/model_container_runner_cuda.h
index 654da9da036..954da3c95fb 100644
--- a/torch/csrc/inductor/aoti_runner/model_container_runner_cuda.h
+++ b/torch/csrc/inductor/aoti_runner/model_container_runner_cuda.h
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #pragma once
 
 #include <c10/cuda/CUDAStream.h>
diff --git a/torch/csrc/inductor/aoti_runner/model_container_runner_xpu.cpp b/torch/csrc/inductor/aoti_runner/model_container_runner_xpu.cpp
index 97dcd1c5913..57bd67968e2 100644
--- a/torch/csrc/inductor/aoti_runner/model_container_runner_xpu.cpp
+++ b/torch/csrc/inductor/aoti_runner/model_container_runner_xpu.cpp
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #include <torch/csrc/inductor/aoti_runner/model_container_runner_xpu.h>
 
 namespace torch::inductor {
diff --git a/torch/csrc/inductor/aoti_runner/model_container_runner_xpu.h b/torch/csrc/inductor/aoti_runner/model_container_runner_xpu.h
index 563d2047f6c..9aa36225f3e 100644
--- a/torch/csrc/inductor/aoti_runner/model_container_runner_xpu.h
+++ b/torch/csrc/inductor/aoti_runner/model_container_runner_xpu.h
@@ -1,4 +1,6 @@
-#if !defined(C10_MOBILE) && !defined(ANDROID)
+// Chaquopy
+// #if !defined(C10_MOBILE) && !defined(ANDROID)
+#ifndef C10_MOBILE
 #pragma once
 
 #include <c10/xpu/XPUStream.h>
diff --git a/torch/lib/libshm/manager.cpp b/torch/lib/libshm/manager.cpp
index 6aa70d8c211..4ef34873977 100644
--- a/torch/lib/libshm/manager.cpp
+++ b/torch/lib/libshm/manager.cpp
@@ -70,7 +70,8 @@ void print_init_message(std::string_view message) {
 }
 
 bool object_exists(const char* name) {
-  int fd = shm_open(name, O_RDONLY, 0);
+  // int fd = shm_open(name, O_RDONLY, 0);
+  int fd = -1;   // Chaquopy: removed shm_open, which isn't supported on Android.
   if (fd >= 0) {
     close(fd);
     return true;
@@ -180,7 +181,8 @@ int main(int argc, char* argv[]) {
 
   for (auto& obj_name : used_objects) {
     DEBUG("freeing %s", obj_name.c_str());
-    shm_unlink(obj_name.c_str());
+    // shm_unlink(obj_name.c_str());
+    // Chaquopy: removed shm_unlink, which isn't supported on Android.
   }
 
   // Clean up file descriptors
