diff --git a/c10/test/core/Scalar_test.cpp b/c10/test/core/Scalar_test.cpp
index d87cc051660..2fb936216d8 100644
--- a/c10/test/core/Scalar_test.cpp
+++ b/c10/test/core/Scalar_test.cpp
@@ -48,8 +48,8 @@ TEST(ScalarTest, Equality) {
   ASSERT_FALSE(Scalar(-1).equal(0xFFFFFFFFFFFFFFFF));
 }
 
-TEST(ScalarTest, LongsAndLongLongs) {
-  Scalar longOne = 1L;
-  Scalar longlongOne = 1LL;
-  ASSERT_EQ(longOne.toInt(), longlongOne.toInt());
-}
+// TEST(ScalarTest, LongsAndLongLongs) {
+//   Scalar longOne = 1L;
+//   Scalar longlongOne = 1LL;
+//   ASSERT_EQ(longOne.toInt(), longlongOne.toInt());
+// }
