diff --git a/CMakeLists.txt b/CMakeLists.txt
index 404fc12..5594dba 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -30,7 +30,8 @@ else()
         set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
     elseif(UNIX)
         # Linux/Unix: add -fPIC and disable -Werror=stringop-overflow
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC -Wno-error=stringop-overflow")
+        # set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC -Wno-error=stringop-overflow")
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
     endif()
 endif()
 
diff --git a/setup.py b/setup.py
index 0f4a9f2..477a76f 100644
--- a/setup.py
+++ b/setup.py
@@ -39,13 +39,14 @@ class CMakeBuild(build_ext):
         cmake_args = [
             "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY={}".format(extdir),
             "-DCMAKE_RUNTIME_OUTPUT_DIRECTORY=$<1:{}>".format(extdir),
-            "-DCMAKE_OSX_ARCHITECTURES=arm64;x86_64",
+            # "-DCMAKE_OSX_ARCHITECTURES=arm64;x86_64",
             # "-DPYTHON_EXECUTABLE={}".format(sys.executable),
             "-DCMAKE_BUILD_TYPE={}".format(cfg),  # not used on MSVC, but no harm
             "-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded",
             "-DOPENVDB_CORE_SHARED=OFF",
             "-DTBB_TEST=OFF",
-            f"-DCMAKE_CXX_FLAGS=-fPIC {'-static-libgcc -static-libstdc++' if system == 'linux' else '/MT /EHsc' if system == 'windows' else ''}"
+            # f"-DCMAKE_CXX_FLAGS=-fPIC {'-static-libgcc -static-libstdc++' if system == 'linux' else '/MT /EHsc' if system == 'windows' else ''}"
+            "-DCMAKE_CXX_FLAGS=-O3 -g0 -Wall"
         ]
         
         build_args = []
