diff --git a/setup.py b/setup.py
index c1d96ca..2e399cd 100644
--- a/setup.py
+++ b/setup.py
@@ -18,7 +18,8 @@ with open(Path(__file__).parent / "extern" / "nlopt" / "CMakeLists.txt") as f:
 
 
 setup(
-    version=version,
+    # version=version,
+    version="2.7.1",
     install_requires=["numpy >=1.18.5"],
     ext_modules=[NLOptBuildExtension("nlopt._nlopt", version)],
     cmdclass={"build_ext": NLOptBuild},
