diff --git a/setup_build.py b/setup_build.py
index 853216e2..6d0b3b1f 100644
--- a/setup_build.py
+++ b/setup_build.py
@@ -86,7 +86,6 @@ class h5py_build_ext(build_ext):
         This is the point at which custom directories, MPI options, etc.
         enter the build process.
         """
-        import numpy
 
         settings = COMPILER_SETTINGS.copy()
 
@@ -99,12 +98,8 @@ class h5py_build_ext(build_ext):
             settings['library_dirs'].extend(config.msmpi_lib_dirs)
             settings['libraries'].append('msmpi')
 
-        try:
-            numpy_includes = numpy.get_include()
-        except AttributeError:
-            # if numpy is not installed get the headers from the .egg directory
-            import numpy.core
-            numpy_includes = os.path.join(os.path.dirname(numpy.core.__file__), 'include')
+        # Chaquopy
+        numpy_includes = os.path.abspath("../requirements/numpy/core/include")
 
         settings['include_dirs'] += [numpy_includes]
         if config.mpi:
@@ -127,9 +122,11 @@ class h5py_build_ext(build_ext):
 
         from Cython import __version__ as cython_version
         from Cython.Build import cythonize
-        import numpy
 
-        complex256_support = hasattr(numpy, 'complex256')
+        # complex256_support = hasattr(numpy, 'complex256')
+        # Chaquopy: was `hasattr(numpy, 'complex256')`: erring on the side of caution.
+        # TODO: see if there's another way to detect this.
+        complex256_support = False
 
         # This allows ccache to recognise the files when pip builds in a temp
         # directory. It speeds up repeatedly running tests through tox with
diff --git a/setup_configure.py b/setup_configure.py
index 3c5fe0db..082d5824 100644
--- a/setup_configure.py
+++ b/setup_configure.py
@@ -122,10 +122,14 @@ class BuildConfig:
 
         Returns (include_dirs, lib_dirs, define_macros)
         """
-        hdf5 = os.environ.get('HDF5_DIR')
-        hdf5_includedir = os.environ.get('HDF5_INCLUDEDIR')
-        hdf5_libdir = os.environ.get('HDF5_LIBDIR')
-        hdf5_pkgconfig_name = os.environ.get('HDF5_PKGCONFIG_NAME')
+        # hdf5 = os.environ.get('HDF5_DIR')
+        # hdf5_includedir = os.environ.get('HDF5_INCLUDEDIR')
+        # hdf5_libdir = os.environ.get('HDF5_LIBDIR')
+        # hdf5_pkgconfig_name = os.environ.get('HDF5_PKGCONFIG_NAME')
+        hdf5 = op.abspath("../requirements/chaquopy")
+        hdf5_includedir = op.join(hdf5, "include")
+        hdf5_libdir = op.join(hdf5, "lib")
+        hdf5_pkgconfig_name = op.join(hdf5, "lib/libhdf5.settings")
 
         if sum([
             bool(hdf5_includedir or hdf5_libdir),
