diff --git a/pyproject.toml b/pyproject.toml
index d181d5bc..b94de8f4 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,7 +1,7 @@
 [build-system]
 requires = [
     "Cython >=0.29.31,<4",
-    "numpy >=2.0.0, <3",
+    "numpy >=1.23.3, <3",
     "pkgconfig",
     "setuptools >=61",
 ]
diff --git a/setup_build.py b/setup_build.py
index 853216e2..9c755810 100644
--- a/setup_build.py
+++ b/setup_build.py
@@ -86,7 +86,8 @@ class h5py_build_ext(build_ext):
         This is the point at which custom directories, MPI options, etc.
         enter the build process.
         """
-        import numpy
+        # todo xiao
+        # import numpy
 
         settings = COMPILER_SETTINGS.copy()
 
@@ -99,12 +100,16 @@ class h5py_build_ext(build_ext):
             settings['library_dirs'].extend(config.msmpi_lib_dirs)
             settings['libraries'].append('msmpi')
 
-        try:
-            numpy_includes = numpy.get_include()
-        except AttributeError:
-            # if numpy is not installed get the headers from the .egg directory
-            import numpy.core
-            numpy_includes = os.path.join(os.path.dirname(numpy.core.__file__), 'include')
+        # todo xiao
+        # try:
+        #     numpy_includes = numpy.get_include()
+        # except AttributeError:
+        #     # if numpy is not installed get the headers from the .egg directory
+        #     import numpy.core
+        #     numpy_includes = os.path.join(os.path.dirname(numpy.core.__file__), 'include')
+        
+        # Chaquopy
+        numpy_includes = os.path.abspath("../requirements/numpy/core/include")
 
         settings['include_dirs'] += [numpy_includes]
         if config.mpi:
@@ -127,9 +132,13 @@ class h5py_build_ext(build_ext):
 
         from Cython import __version__ as cython_version
         from Cython.Build import cythonize
-        import numpy
+        
+        # todo xiao
+        # import numpy
 
-        complex256_support = hasattr(numpy, 'complex256')
+        # todo xiao
+        # complex256_support = hasattr(numpy, 'complex256')
+        complex256_support = False
 
         # This allows ccache to recognise the files when pip builds in a temp
         # directory. It speeds up repeatedly running tests through tox with
@@ -137,6 +146,9 @@ class h5py_build_ext(build_ext):
         # ccache is not in use.
         os.environ['CCACHE_BASEDIR'] = op.dirname(op.abspath(__file__))
         os.environ['CCACHE_NOHASHDIR'] = '1'
+        
+        # todo xiao
+        os.environ['HDF5_DIR'] = '/home/ubuntu/workspace/pkg/hdf5'
 
         # Get configuration from environment variables
         config = BuildConfig.from_env()
diff --git a/setup_configure.py b/setup_configure.py
index 3c5fe0db..d5b41371 100644
--- a/setup_configure.py
+++ b/setup_configure.py
@@ -89,6 +89,10 @@ class BuildConfig:
         h5_version_s = os.environ.get('HDF5_VERSION')
         h5py_ros3 = os.environ.get('H5PY_ROS3')
         h5py_direct_vfd = os.environ.get('H5PY_DIRECT_VFD')
+        
+        print(f"-------------------from_env---h5_version_s = {h5_version_s}")
+        print(f"-------------------from_env---h5py_ros3 = {h5py_ros3}")
+        print(f"-------------------from_env---h5py_direct_vfd = {h5py_direct_vfd}")
 
         if h5_version_s and not mpi and h5py_ros3 and h5py_direct_vfd:
             # if we know config, don't use wrapper, it may not be supported
@@ -126,6 +130,11 @@ class BuildConfig:
         hdf5_includedir = os.environ.get('HDF5_INCLUDEDIR')
         hdf5_libdir = os.environ.get('HDF5_LIBDIR')
         hdf5_pkgconfig_name = os.environ.get('HDF5_PKGCONFIG_NAME')
+        
+        print(f"---------------------------------_find_hdf5_compiler_settings---hdf5 = {hdf5}")
+        print(f"---------------------------------_find_hdf5_compiler_settings---hdf5_includedir = {hdf5_includedir}")
+        print(f"---------------------------------_find_hdf5_compiler_settings---hdf5_libdir = {hdf5_libdir}")
+        print(f"---------------------------------_find_hdf5_compiler_settings---hdf5_pkgconfig_name = {hdf5_pkgconfig_name}")
 
         if sum([
             bool(hdf5_includedir or hdf5_libdir),
