Submodule opencv contains modified content
diff --git a/opencv/CMakeLists.txt b/opencv/CMakeLists.txt
index e60406fbe2..c0ea4ebd44 100644
--- a/opencv/CMakeLists.txt
+++ b/opencv/CMakeLists.txt
@@ -427,7 +427,7 @@ OCV_OPTION(WITH_GPHOTO2 "Include gPhoto2 library support" OFF
   VISIBLE_IF UNIX AND NOT ANDROID AND NOT IOS AND NOT XROS
   VERIFY HAVE_GPHOTO2)
 OCV_OPTION(WITH_LAPACK "Include Lapack library support" (NOT CV_DISABLE_OPTIMIZATION)
-  VISIBLE_IF NOT ANDROID AND NOT IOS AND NOT XROS
+  VISIBLE_IF NOT IOS AND NOT XROS # Chaquopy: removed ANDROID exclusion
   VERIFY HAVE_LAPACK)
 OCV_OPTION(WITH_ITT "Include Intel ITT support" ON
   VISIBLE_IF NOT APPLE_FRAMEWORK
diff --git a/opencv/cmake/OpenCVDetectPython.cmake b/opencv/cmake/OpenCVDetectPython.cmake
index a23fba6e5a..0141af26f9 100644
--- a/opencv/cmake/OpenCVDetectPython.cmake
+++ b/opencv/cmake/OpenCVDetectPython.cmake
@@ -123,7 +123,7 @@ if(NOT ${found})
   if(_found)
     set(_version_major_minor "${_version_major}.${_version_minor}")
 
-    if(NOT ANDROID AND NOT APPLE_FRAMEWORK)
+    if(NOT APPLE_FRAMEWORK)  # Chaquopy: removed ANDROID exclusion
       ocv_check_environment_variables(${library_env} ${include_dir_env})
       if(NOT ${${library_env}} STREQUAL "")
           set(PYTHON_LIBRARY "${${library_env}}")
@@ -144,9 +144,9 @@ if(NOT ${found})
         # Copy outputs
         set(_libs_found ${PYTHONLIBS_FOUND})
         set(_libraries ${PYTHON_LIBRARIES})
-        set(_include_path ${PYTHON_INCLUDE_PATH})
+        set(_include_path ${PYTHON_INCLUDE_DIR})  # Chaquopy: replaced PYTHON_INCLUDE_PATH and
         set(_include_dirs ${PYTHON_INCLUDE_DIRS})
-        set(_debug_libraries ${PYTHON_DEBUG_LIBRARIES})
+        set(_include_dirs ${PYTHON_INCLUDE_DIR})  # PYTHON_INCLUDE_DIRS, which were always empty.
         set(_libs_version_string ${PYTHONLIBS_VERSION_STRING})
         set(_debug_library ${PYTHON_DEBUG_LIBRARY})
         set(_library ${PYTHON_LIBRARY})
@@ -175,7 +175,7 @@ if(NOT ${found})
       endif()
     endif()
 
-    if(NOT ANDROID AND NOT IOS AND NOT XROS)
+    if(NOT IOS AND NOT XROS) # Chaquopy: removed ANDROID exclusion
       if(CMAKE_HOST_UNIX)
         execute_process(COMMAND ${_executable} -c "from sysconfig import *; print(get_path('purelib'))"
                         RESULT_VARIABLE _cvpy_process
@@ -240,7 +240,7 @@ if(NOT ${found})
                           OUTPUT_STRIP_TRAILING_WHITESPACE)
         endif()
       endif()
-    endif(NOT ANDROID AND NOT IOS AND NOT XROS)
+    endif(NOT IOS AND NOT XROS)
   endif()
 
   # Export return values
diff --git a/opencv/modules/python/CMakeLists.txt b/opencv/modules/python/CMakeLists.txt
index 93eab8c94d..ded0365e93 100644
--- a/opencv/modules/python/CMakeLists.txt
+++ b/opencv/modules/python/CMakeLists.txt
@@ -3,7 +3,7 @@
 # ----------------------------------------------------------------------------
 if(DEFINED OPENCV_INITIAL_PASS)  # OpenCV build
 
-if(ANDROID OR APPLE_FRAMEWORK OR WINRT)
+if(APPLE_FRAMEWORK OR WINRT) # Chaquopy: removed ANDROID exclusion
   ocv_module_disable_(python2)
   ocv_module_disable_(python3)
   return()
diff --git a/setup.py b/setup.py
index 1dfab2f..43e1e32 100644
--- a/setup.py
+++ b/setup.py
@@ -17,8 +17,10 @@ def main():
     is_CI_build = True if CI_BUILD == "1" else False
     cmake_source_dir = "opencv"
     minimum_supported_numpy = "1.13.3"
-    build_contrib = get_build_env_var_by_name("contrib")
-    build_headless = get_build_env_var_by_name("headless")
+    # Chaquopy
+    build_contrib = "contrib" in os.environ["PKG_NAME"]
+    build_headless = "headless" in os.environ["PKG_NAME"]
+    
     build_java = "ON" if get_build_env_var_by_name("java") else "OFF"
     build_rolling = get_build_env_var_by_name("rolling")
 
@@ -36,17 +38,37 @@ def main():
     ]
 
     python_version = cmaker.CMaker.get_python_version()
-    python_lib_path = cmaker.CMaker.get_python_library(python_version) or ""
-    # HACK: For Scikit-build 0.17.3 and newer that returns None or empty sptring for PYTHON_LIBRARY in manylinux2014
-    # A small release related to PYTHON_LIBRARY handling changes in 0.17.2; scikit-build 0.17.3 returns an empty string from get_python_library if no Python library is present (like on manylinux), where 0.17.2 returned None, and previous versions returned a non-existent path. Note that adding REQUIRED to find_package(PythonLibs will fail, but it is incorrect (you must not link to libPython.so) and was really just injecting a non-existent path before.
-    # TODO: Remove the hack when the issue is handled correctly in main OpenCV CMake.
-    if python_lib_path == "":
-        python_lib_path = "libpython%sm.a" % python_version
-    python_lib_path = python_lib_path.replace("\\", "/")
-
-    python_include_dir = cmaker.CMaker.get_python_include_dir(python_version).replace(
-        "\\", "/"
-    )
+    # python_lib_path = cmaker.CMaker.get_python_library(python_version) or ""
+    # # HACK: For Scikit-build 0.17.3 and newer that returns None or empty sptring for PYTHON_LIBRARY in manylinux2014
+    # # A small release related to PYTHON_LIBRARY handling changes in 0.17.2; scikit-build 0.17.3 returns an empty string from get_python_library if no Python library is present (like on manylinux), where 0.17.2 returned None, and previous versions returned a non-existent path. Note that adding REQUIRED to find_package(PythonLibs will fail, but it is incorrect (you must not link to libPython.so) and was really just injecting a non-existent path before.
+    # # TODO: Remove the hack when the issue is handled correctly in main OpenCV CMake.
+    # if python_lib_path == "":
+    #     python_lib_path = "libpython%sm.a" % python_version
+    # python_lib_path = python_lib_path.replace("\\", "/")
+
+    # python_include_dir = cmaker.CMaker.get_python_include_dir(python_version).replace(
+    #     "\\", "/"
+    # )
+    
+    # Chaquopy
+    python_include_dir = None
+    in_dirafter = False
+    for word in os.environ["CFLAGS"].split():
+        if word == "-idirafter":
+            in_dirafter = True
+        elif in_dirafter:
+            in_dirafter = False
+            if "include/python" in word:
+                python_include_dir = word
+    assert python_include_dir, os.environ["CFLAGS"]
+    
+    # Chaquopy
+    python_lib_dir = os.path.abspath(python_include_dir + "/../../lib")
+    python_lib_path = None
+    for word in os.environ["LDFLAGS"].split():
+        if word.startswith("-lpython"):
+            python_lib_path = f"{python_lib_dir}/lib{word[2:]}.so"
+    assert python_lib_path, os.environ["LDFLAGS"]
 
     if os.path.exists(".git"):
         import pip._internal.vcs.git as git
@@ -79,7 +101,7 @@ def main():
     )
 
     # https://stackoverflow.com/questions/1405913/python-32bit-or-64bit-mode
-    is64 = sys.maxsize > 2 ** 32
+    is64 = "64" in os.environ["CHAQUOPY_ABI"]  # Chaquopy: was sys.maxsize > 2 ** 32
 
     package_name = "opencv-python"
 
@@ -91,6 +113,9 @@ def main():
 
     if build_headless and not build_contrib:
         package_name = "opencv-python-headless"
+    
+    # Chaquopy
+    package_name = os.environ["PKG_NAME"]
 
     if build_rolling:
         package_name += "-rolling"
@@ -137,7 +162,8 @@ def main():
         [ r"python/cv2/py.typed" ] if sys.version_info >= (3, 6) else []
         ,
         "cv2.data": [  # OPENCV_OTHER_INSTALL_PATH
-            ("etc" if os.name == "nt" else "share/opencv4") + r"/haarcascades/.*\.xml"
+            # Chaquopy: see opencv/cmake/OpenCVInstallLayout.cmake
+            "sdk/etc" + r"/haarcascades/.*\.xml"
         ],
         "cv2.gapi": [
             "python/cv2" + r"/gapi/.*\.py"
@@ -212,7 +238,34 @@ def main():
             else []
         )
     )
-
+    print("-------------------------------------opencv python---setup---1")
+    # Chaquopy
+    # py_major_ver = sys.version_info[0]
+    cmake_args += [
+        # Enable CMake's cross-compilation mode.
+        "-DCMAKE_TOOLCHAIN_FILE=" + os.path.abspath("../chaquopy.toolchain.cmake"),
+        
+        # Unnecessary, and requires an old version of the SDK.
+        "-DBUILD_ANDROID_PROJECTS=OFF", "-DBUILD_ANDROID_EXAMPLES=OFF",
+        
+        # External dependencies which we've built as wheels (listed in meta.yaml).
+        "-DBUILD_JPEG=OFF", "-DBUILD_PNG=OFF",
+        
+        # External dependencies which we can do without for now.
+        "-DWITH_TIFF=OFF", "-DWITH_JASPER=OFF", "-DWITH_OPENEXR=OFF", "-DWITH_WEBP=OFF",
+        
+        # These are only available for x86, so disable for consistency between ABIs.
+        "-DWITH_IPP=OFF", "-DWITH_ITT=OFF",
+        
+        # Python native modules aren't usually linked against libpython on Linux, but our
+        # configuration requires it (see opencv/modules/python/common.cmake).
+        "-DOPENCV_FORCE_PYTHON_LIBS=ON",
+        
+        # "-DPYTHON{}_NUMPY_INCLUDE_DIRS={}".format(
+        #     py_major_ver, os.path.abspath("../requirements/numpy/core/include")),
+    ]
+    
+    print("-------------------------------------opencv python---setup---2")
     if build_headless:
         # it seems that cocoa cannot be disabled so on macOS the package is not truly headless
         cmake_args.append("-DWITH_WIN32UI=OFF")
@@ -225,7 +278,7 @@ def main():
 
     if sys.platform.startswith("linux") and not is64 and "bdist_wheel" in sys.argv:
         subprocess.check_call("patch -p0 < patches/patchOpenEXR", shell=True)
-
+    print("-------------------------------------opencv python---setup---3")
     # OS-specific components during CI builds
     if is_CI_build:
 
@@ -261,12 +314,12 @@ def main():
             cmake_args.append("-DWITH_V4L=ON")
             cmake_args.append("-DWITH_LAPACK=ON")
             cmake_args.append("-DENABLE_PRECOMPILED_HEADERS=OFF")
-
+    print("-------------------------------------opencv python---setup---4")
     # works via side effect
     RearrangeCMakeOutput(
         rearrange_cmake_output_data, files_outside_package_dir, package_data.keys()
     )
-
+    print(f"-------------------------------------opencv python---setup---5---cmake_args = {cmake_args}")
     setup(
         name=package_name,
         version=package_version,
@@ -279,7 +332,8 @@ def main():
         package_data=package_data,
         maintainer="OpenCV Team",
         ext_modules=EmptyListWithLength(),
-        install_requires=install_requires,
+        # Chaquopy: install_requires numpy is handled by meta.yaml.
+        # install_requires=install_requires,
         python_requires=">=3.6",
         classifiers=[
             "Development Status :: 5 - Production/Stable",
@@ -336,10 +390,11 @@ class RearrangeCMakeOutput:
     files_outside_package = None
 
     def __init__(self, package_paths_re, files_outside_package, packages):
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---1")
         cls = self.__class__
         assert not cls.wraps._classify_installed_files, "Singleton object"
         import skbuild.setuptools_wrap
-
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---2")
         cls._setuptools_wrap = skbuild.setuptools_wrap
         cls.wraps._classify_installed_files = (
             cls._setuptools_wrap._classify_installed_files
@@ -347,18 +402,20 @@ class RearrangeCMakeOutput:
         cls._setuptools_wrap._classify_installed_files = (
             self._classify_installed_files_override
         )
-
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---3")
         cls.package_paths_re = package_paths_re
         cls.files_outside_package = files_outside_package
         cls.packages = packages
 
     def __del__(self):
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---4")
         cls = self.__class__
         cls._setuptools_wrap._classify_installed_files = (
             cls.wraps._classify_installed_files
         )
         cls.wraps._classify_installed_files = None
         cls._setuptools_wrap = None
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---5")
 
     def _classify_installed_files_override(
         self,
@@ -382,7 +439,7 @@ class RearrangeCMakeOutput:
                     subpackage\
                         etc.
         """
-
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---6")
         cls = self.__class__
 
         # 'relpath'/'reldir' = relative to CMAKE_INSTALL_DIR/cmake_install_dir
@@ -399,7 +456,7 @@ class RearrangeCMakeOutput:
         relpaths_zip = list(zip(fslash_install_relpaths, install_relpaths))
 
         final_install_relpaths = []
-
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---7")
         print("Copying files from CMake output")
 
         # add lines from the old __init__.py file to the config file
@@ -411,7 +468,7 @@ class RearrangeCMakeOutput:
                                  % (sys.version_info[0], sys.version_info[1]))
         if not os.path.exists(config_py):
             config_py = os.path.join(cmake_install_dir, 'python', 'cv2', 'config-%s.py' % sys.version_info[0])
-
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---8")
         with open(config_py, 'w') as opencv_init_config:
             opencv_init_config.write(custom_init_data)
 
@@ -425,7 +482,7 @@ class RearrangeCMakeOutput:
                         hide_listing=False,
                     )
                     final_install_relpaths.append(os.path.join("cv2", target_rel_path))
-
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---9")
         del install_relpaths, fslash_install_relpaths
 
         for package_name, relpaths_re in cls.package_paths_re.items():
@@ -453,7 +510,7 @@ class RearrangeCMakeOutput:
                     if not found and "gapi" not in relpath_re:
                         raise Exception("Not found: '%s'" % relpath_re)
                 del r, found
-
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---10")
         del relpaths_zip
 
         print("Copying files from non-default sourcetree locations")
@@ -473,11 +530,11 @@ class RearrangeCMakeOutput:
                     hide_listing=False,
                 )
                 final_install_relpaths.append(new_install_relpath)
-
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---11")
         final_install_paths = [
             os.path.join(cmake_install_dir, p) for p in final_install_relpaths
         ]
-
+        print("-------------------------------------opencv python---RearrangeCMakeOutput---12")
         return (cls.wraps._classify_installed_files)(
             final_install_paths,
             package_data,
@@ -495,10 +552,11 @@ class RearrangeCMakeOutput:
 
 def get_and_set_info(contrib, headless, rolling, ci_build):
     # cv2/version.py should be generated by running find_version.py
+    print("-------------------------------------opencv python---get_and_set_info---1")
     version = {}
     here = os.path.abspath(os.path.dirname(__file__))
     version_file = os.path.join(here, "cv2", "version.py")
-
+    print("-------------------------------------opencv python---get_and_set_info---2")
     # generate a fresh version.py always when Git repository exists
     # (in sdists the version.py file already exists)
     if os.path.exists(".git"):
@@ -506,27 +564,28 @@ def get_and_set_info(contrib, headless, rolling, ci_build):
         sys.argv = ["", str(contrib), str(headless), str(rolling), str(ci_build)]
         runpy.run_path("find_version.py", run_name="__main__")
         sys.argv = old_args
-
+    print("-------------------------------------opencv python---get_and_set_info---3")
     with open(version_file) as fp:
         exec(fp.read(), version)
-
+    ver = version["opencv_version"]
+    print(f"--------------------get_and_set_info---version[opencv_version] = {ver}")
     return version["opencv_version"], version["contrib"], version["headless"], version["rolling"]
 
 
 def get_build_env_var_by_name(flag_name):
     flag_set = False
-
+    print("-------------------------------------opencv python---get_build_env_var_by_name---1")
     try:
         flag_set = bool(int(os.getenv("ENABLE_" + flag_name.upper(), None)))
     except Exception:
         pass
-
+    print("-------------------------------------opencv python---get_build_env_var_by_name---2")
     if not flag_set:
         try:
             flag_set = bool(int(open(flag_name + ".enabled").read(1)))
         except Exception:
             pass
-
+    print("-------------------------------------opencv python---get_build_env_var_by_name---3")
     return flag_set
 
 
